File: src/main/java/utils/SelfHealingLocator.java
package utils;

import com.microsoft.playwright.Locator;
import com.microsoft.playwright.Page;
import java.util.ArrayList;
import java.util.List;

/**
 * Basic standalone implementation of SelfHealingLocator.
 * This mirrors the interface in your main test workspace.
 */
public class SelfHealingLocator {

    @FunctionalInterface
    public interface LocatorAction {
        void perform(Locator locator);
    }

    private static void performWithHealing(Page page, String key, String defaultSelector, LocatorAction action) {
        try {
            // Set 10-second timeout for primary locator attempt to allow async page transitions to complete
            page.setDefaultTimeout(10000);
            action.perform(page.locator(defaultSelector));
            page.setDefaultTimeout(30000); // restore default timeout
        } catch (com.microsoft.playwright.PlaywrightException e) {
            page.setDefaultTimeout(30000); // restore default timeout on exception
            String errorMsg = e.getMessage();
            System.err.println("[Self-Healing] Error interacting with element '" + key + "' using selector '"
                    + defaultSelector + "': " + errorMsg);

            // 2. Strict Mode Violation Healing
            if (errorMsg != null && errorMsg.contains("strict mode violation")) {
                System.out.println(
                        "[Self-Healing] Strict mode violation detected. Attempting to find a visible and enabled matching element...");
                Locator locator = page.locator(defaultSelector);
                try {
                    int count = locator.count();
                    for (int i = 0; i < count; i++) {
                        Locator item = locator.nth(i);
                        if (item.isVisible() && item.isEnabled()) {
                            System.out.println(
                                    "[Self-Healing] Success! Resolved strict mode violation by interacting with match index "
                                            + i);
                            action.perform(item);
                            return;
                        }
                    }
                } catch (Exception ex) {
                    System.err.println("[Self-Healing] Failed during strict mode item iteration: " + ex.getMessage());
                }

                // Fallback to the first matching element
                System.out.println("[Self-Healing] Falling back to first matching element.");
                try {
                    action.perform(locator.first());
                    return;
                } catch (Exception ex) {
                    System.err.println("[Self-Healing] Fallback to first element failed: " + ex.getMessage());
                }
            }

            // 3. Element Not Found / Timeout / Locator Changed Healing
            System.out.println("[Self-Healing] Attempting text-based recovery healing...");
            String extractedText = extractText(defaultSelector);
            if (extractedText != null && !extractedText.isEmpty()) {
                String cleanedText = extractedText.replaceAll("\\s+", " ").trim();
                List<String> fallbacks = new ArrayList<>();
                
                // Full text fallbacks
                fallbacks.add("text=" + extractedText);
                fallbacks.add("xpath=//*[contains(text(), '" + extractedText + "')]");
                fallbacks.add("xpath=//*[contains(normalize-space(), '" + extractedText + "')]");

                String[] words = cleanedText.split("\\s+");
                // 1-word fallback (e.g. "Brands")
                if (words.length >= 1 && words[0].length() >= 3) {
                    String firstWord = words[0];
                    if (!firstWord.equalsIgnoreCase(extractedText)) {
                        fallbacks.add("text=" + firstWord);
                        fallbacks.add("xpath=//*[contains(normalize-space(), '" + firstWord + "')]");
                        fallbacks.add("xpath=//button[contains(normalize-space(), '" + firstWord + "')]");
                        fallbacks.add("xpath=//a[contains(normalize-space(), '" + firstWord + "')]");
                        fallbacks.add("xpath=//div[contains(normalize-space(), '" + firstWord + "')]");
                    }
                }
                // 2-word fallback (e.g. "Brands Controlled")
                if (words.length >= 2) {
                    String twoWords = words[0] + " " + words[1];
                    if (!twoWords.equalsIgnoreCase(extractedText)) {
                        fallbacks.add("text=" + twoWords);
                        fallbacks.add("xpath=//*[contains(normalize-space(), '" + twoWords + "')]");
                    }
                }
                // 3-word fallback (e.g. "Brands Controlled list")
                if (words.length >= 3) {
                    String threeWords = words[0] + " " + words[1] + " " + words[2];
                    if (!threeWords.equalsIgnoreCase(extractedText)) {
                        fallbacks.add("text=" + threeWords);
                        fallbacks.add("xpath=//*[contains(normalize-space(), '" + threeWords + "')]");
                    }
                }

                String cleanAlphaNum = cleanedText.replaceAll("[^a-zA-Z0-9 ]", "").trim().replaceAll("\\s+", " ");
                if (!cleanAlphaNum.isEmpty() && !cleanAlphaNum.equals(extractedText)) {
                    fallbacks.add("text=" + cleanAlphaNum);
                    fallbacks.add("xpath=//*[contains(normalize-space(), '" + cleanAlphaNum + "')]");
                }

                for (String fallback : fallbacks) {
                    try {
                        System.out.println("[Self-Healing] Trying fallback selector: " + fallback);
                        Locator fallbackLocator = page.locator(fallback);
                        int count = fallbackLocator.count();
                        if (count > 0) {
                            for (int i = 0; i < count; i++) {
                                Locator item = fallbackLocator.nth(i);
                                if (item.isVisible() && item.isEnabled()) {
                                    System.out.println("[Self-Healing Success] Successfully healed '" + key
                                            + "' using fallback selector: '" + fallback + "'");
                                    persistHealedSelector(key, fallback);
                                    action.perform(item);
                                    return;
                                }
                            }
                        }
                    } catch (Exception ex) {
                        // ignore and try next fallback
                    }
                }
            }

            // Diagnostic summary if healing attempts fail
            System.err.println("[Self-Healing Warning] Unable to find or interact with '" + key + "' (selector: " + defaultSelector + ").");
            System.err.println("[Self-Healing Diagnostic] Current Page URL: " + page.url());

            // Re-throw the original exception if all healing attempts fail
            throw e;
        }
    }

    private static String extractText(String selector) {
        if (selector == null)
            return null;
        int firstQuote = selector.indexOf('\'');
        if (firstQuote != -1) {
            int secondQuote = selector.indexOf('\'', firstQuote + 1);
            if (secondQuote != -1) {
                return selector.substring(firstQuote + 1, secondQuote);
            }
        }
        int firstDblQuote = selector.indexOf('"');
        if (firstDblQuote != -1) {
            int secondDblQuote = selector.indexOf('"', firstDblQuote + 1);
            if (secondDblQuote != -1) {
                return selector.substring(firstDblQuote + 1, secondDblQuote);
            }
        }
        return null;
    }

    public static void click(Page page, String key, String defaultSelector) {
        performWithHealing(page, key, defaultSelector, loc -> {
            try {
                loc.click();
            } catch (Exception e) {
                loc.click(new com.microsoft.playwright.Locator.ClickOptions().setForce(true));
            }
        });
    }

    public static void fill(Page page, String key, String defaultSelector, String value) {
        performWithHealing(page, key, defaultSelector, loc -> loc.fill(value));
    }

    public static void selectOption(Page page, String key, String defaultSelector, String value) {
        performWithHealing(page, key, defaultSelector, loc -> loc.selectOption(value));
    }

    public static void check(Page page, String key, String defaultSelector) {
        performWithHealing(page, key, defaultSelector, Locator::check);
    }

    public static void uncheck(Page page, String key, String defaultSelector) {
        performWithHealing(page, key, defaultSelector, Locator::uncheck);
    }

    public static void press(Page page, String key, String defaultSelector, String value) {
        performWithHealing(page, key, defaultSelector, loc -> loc.press(value));
    }

    public static void hover(Page page, String key, String defaultSelector) {
        // Extract human-readable text from selector for JS fallback
        String hoverText = defaultSelector;
        if (hoverText.startsWith("text=")) {
            hoverText = hoverText.substring(5).trim();
        } else if (hoverText.contains("normalize-space()='") || hoverText.contains("contains(.,")) {
            // Extract text between quotes
            int q1 = hoverText.lastIndexOf("'");
            int q0 = hoverText.lastIndexOf("'", q1 - 1);
            if (q0 >= 0 && q1 > q0) hoverText = hoverText.substring(q0 + 1, q1);
        }
        final String textForJS = hoverText;

        performWithHealing(page, key, defaultSelector, loc -> {
            // Strategy 1: Standard Playwright hover (scroll + hover)
            try {
                loc.first().scrollIntoViewIfNeeded();
                loc.first().hover();
                return;
            } catch (Exception ignored) {}

            // Strategy 2: Force hover (bypasses visibility check)
            try {
                loc.first().hover(new com.microsoft.playwright.Locator.HoverOptions().setForce(true));
                System.out.println("[Hover] Force hover succeeded for: " + key);
                return;
            } catch (Exception ignored) {}

            // Strategy 3: JS mouseover — works for collapsed nav/sidebars
            // Finds the element by text and dispatches mouseover + mouseenter events
            try {
                String js = "(text) => {" +
                    "var els = Array.from(document.querySelectorAll('button,a,li,[role]'));" +
                    "var el = els.find(function(e){ return e.textContent && e.textContent.trim().indexOf(text) !== -1; });" +
                    "if (el) {" +
                    "  el.dispatchEvent(new MouseEvent('mouseover', {bubbles:true,cancelable:true}));" +
                    "  el.dispatchEvent(new MouseEvent('mouseenter', {bubbles:true,cancelable:true}));" +
                    "}" +
                    "}";
                page.evaluate(js, textForJS);
                System.out.println("[Hover] JS mouseover dispatched for: " + key + " (text='" + textForJS + "')");
                // Give the page time to react to the hover event before next action
                page.waitForTimeout(500);
            } catch (Exception e3) {
                System.err.println("[Hover] All hover strategies failed for '" + key + "': " + e3.getMessage());
                throw new RuntimeException("Hover failed for " + key, e3);
            }
        });
    }



    private static void persistHealedSelector(String key, String healedSelector) {
        try {
            String projectRoot = System.getProperty("user.dir");
            java.io.File poDir = new java.io.File(projectRoot, "src/main/java/pageobjects");
            if (!poDir.exists()) {
                return;
            }
            java.io.File[] files = poDir.listFiles((dir, name) -> name.endsWith(".java"));
            if (files == null) return;
            for (java.io.File file : files) {
                String content = new String(java.nio.file.Files.readAllBytes(file.toPath()));
                java.util.regex.Pattern pattern = java.util.regex.Pattern.compile(
                    "(public\\s+static\\s+final\\s+String\\s+" + java.util.regex.Pattern.quote(key) + "\\s*=\\s*\")([^\"]+)(\";)"
                );
                java.util.regex.Matcher matcher = pattern.matcher(content);
                if (matcher.find()) {
                    String newContent = matcher.replaceFirst("$1" + java.util.regex.Matcher.quoteReplacement(healedSelector) + "$3");
                    java.nio.file.Files.write(file.toPath(), newContent.getBytes());
                    System.out.println("[Self-Healing Persistent] Auto-saved healed selector for '" + key + "' in " + file.getName() + " -> \"" + healedSelector + "\"");
                }
            }
        } catch (Exception e) {
            System.err.println("[Self-Healing Warning] Failed to save healed selector: " + e.getMessage());
        }
    }
}

File: typescript-tests/framework/utils/SelfHealingLocator.ts
import { Page, Locator } from '@playwright/test';
import * as fs from 'fs';
import * as path from 'path';

export type LocatorAction = (locator: Locator) => Promise<void>;

export class SelfHealingLocator {

  static async performWithHealing(
    page: Page,
    key: string,
    defaultSelector: string,
    action: LocatorAction
  ): Promise<void> {
    try {
      // 1. Primary locator attempt with a 10s default timeout for transitions
      const originalTimeout = 30000;
      page.setDefaultTimeout(10000);
      await action(page.locator(defaultSelector));
      page.setDefaultTimeout(originalTimeout);
    } catch (err: any) {
      page.setDefaultTimeout(30000);
      const errorMsg: string = err?.message || String(err);
      console.warn(`[Self-Healing] Action failed for '${key}' using '${defaultSelector}': ${errorMsg}`);

      // 2. Strict Mode Violation Healing (multiple matching elements)
      if (errorMsg.includes('strict mode violation')) {
        console.log(`[Self-Healing] Strict mode violation detected for '${key}'. Resolving to first visible match...`);
        const locator = page.locator(defaultSelector);
        try {
          const count = await locator.count();
          for (let i = 0; i < count; i++) {
            const item = locator.nth(i);
            if (await item.isVisible() && await item.isEnabled()) {
              console.log(`[Self-Healing Success] Resolved strict mode for '${key}' using match index ${i}`);
              await action(item);
              return;
            }
          }
        } catch (subErr) {
          console.error(`[Self-Healing] Iteration through strict mode elements failed:`, subErr);
        }

        // Fallback to first element
        try {
          await action(locator.first());
          return;
        } catch (firstErr) {
          console.error(`[Self-Healing] Fallback to first element failed:`, firstErr);
        }
      }

      // 3. Text & Structural Fallbacks
      console.log(`[Self-Healing] Attempting fallback locator healing for '${key}'...`);
      const extractedText = this.extractText(defaultSelector);
      const fallbacks: string[] = [];

      if (extractedText && extractedText.trim().length > 0) {
        const cleaned = extractedText.trim();
        fallbacks.push(`text=${cleaned}`);
        fallbacks.push(`xpath=//*[contains(text(), '${cleaned}')]`);
        fallbacks.push(`xpath=//*[contains(normalize-space(), '${cleaned}')]`);

        const words = cleaned.split(/\s+/);
        if (words.length >= 1 && words[0].length >= 3) {
          fallbacks.push(`xpath=//button[contains(normalize-space(), '${words[0]}')]`);
          fallbacks.push(`xpath=//a[contains(normalize-space(), '${words[0]}')]`);
          fallbacks.push(`xpath=//input[contains(@id, '${words[0].toLowerCase()}')]`);
          fallbacks.push(`xpath=//input[contains(@name, '${words[0].toLowerCase()}')]`);
        }
      }

      // If selector contains id or name, generate fallback candidates
      const idMatch = defaultSelector.match(/#([a-zA-Z0-9_-]+)/);
      if (idMatch) {
        fallbacks.push(`[data-test="${idMatch[1]}"]`);
        fallbacks.push(`[name="${idMatch[1]}"]`);
        fallbacks.push(`[placeholder*="${idMatch[1]}" i]`);
      }

      for (const fallback of fallbacks) {
        try {
          console.log(`[Self-Healing] Trying fallback selector: ${fallback}`);
          const fbLocator = page.locator(fallback);
          const count = await fbLocator.count();
          if (count > 0) {
            for (let i = 0; i < count; i++) {
              const item = fbLocator.nth(i);
              if (await item.isVisible() && await item.isEnabled()) {
                console.log(`[Self-Healing Success] Successfully healed '${key}' using fallback: '${fallback}'`);
                await action(item);
                this.persistHealedSelector(key, fallback);
                return;
              }
            }
          }
        } catch {
          // ignore and proceed to next candidate
        }
      }

      console.error(`[Self-Healing Warning] All healing attempts failed for '${key}' (${defaultSelector})`);
      console.error(`[Self-Healing Diagnostic] Current URL: ${page.url()}`);
      throw err;
    }
  }

  private static extractText(selector: string): string | null {
    if (!selector) return null;
    const match = selector.match(/['"]([^'"]+)['"]/);
    return match ? match[1] : null;
  }

  static async click(page: Page, key: string, defaultSelector: string): Promise<void> {
    await this.performWithHealing(page, key, defaultSelector, async (loc) => {
      try {
        await loc.click({ timeout: 5000 });
      } catch {
        await loc.click({ force: true });
      }
    });
  }

  static async fill(page: Page, key: string, defaultSelector: string, value: string): Promise<void> {
    await this.performWithHealing(page, key, defaultSelector, async (loc) => {
      await loc.fill(value);
    });
  }

  static async selectOption(page: Page, key: string, defaultSelector: string, value: string): Promise<void> {
    await this.performWithHealing(page, key, defaultSelector, async (loc) => {
      await loc.selectOption(value);
    });
  }

  static async check(page: Page, key: string, defaultSelector: string): Promise<void> {
    await this.performWithHealing(page, key, defaultSelector, async (loc) => {
      await loc.check();
    });
  }

  static async uncheck(page: Page, key: string, defaultSelector: string): Promise<void> {
    await this.performWithHealing(page, key, defaultSelector, async (loc) => {
      await loc.uncheck();
    });
  }

  static async press(page: Page, key: string, defaultSelector: string, keyToPress: string): Promise<void> {
    await this.performWithHealing(page, key, defaultSelector, async (loc) => {
      await loc.press(keyToPress);
    });
  }

  static async hover(page: Page, key: string, defaultSelector: string): Promise<void> {
    await this.performWithHealing(page, key, defaultSelector, async (loc) => {
      try {
        await loc.first().scrollIntoViewIfNeeded();
        await loc.first().hover();
      } catch {
        await loc.first().hover({ force: true });
      }
    });
  }

  private static persistHealedSelector(key: string, healedSelector: string): void {
    try {
      const pagesDir = path.resolve(__dirname, '../../pages');
      if (!fs.existsSync(pagesDir)) return;
      const files = fs.readdirSync(pagesDir).filter(f => f.endsWith('.ts'));
      for (const file of files) {
        const fullPath = path.join(pagesDir, file);
        let content = fs.readFileSync(fullPath, 'utf8');
        const regex = new RegExp(`(static\\s+readonly\\s+${key}\\s*=\\s*['"])([^'"]+)(['"];)`);
        if (regex.test(content)) {
          content = content.replace(regex, `$1${healedSelector}$3`);
          fs.writeFileSync(fullPath, content, 'utf8');
          console.log(`[Self-Healing Persistent] Updated selector for '${key}' in ${file} -> "${healedSelector}"`);
        }
      }
    } catch (e: any) {
      console.warn(`[Self-Healing Warning] Could not persist healed selector: ${e?.message}`);
    }
  }
}
